File: 20.Valid-Parentheses/java/Valid_Parentheses.java
import java.util.HashMap;
import java.util.Stack;

public class Valid_Parentheses {
    public static boolean isValid(String s) {
        HashMap var1 = new HashMap();
        var1.put(')', '(');
        var1.put(']', '[');
        var1.put('}', '{');
        Stack stack = new Stack();

        for (char c : s.toCharArray()) {
            if (var1.containsValue(c)) {
                stack.push(c);
            } else {
                if (stack.isEmpty() || stack.pop() != var1.get(c)) {
                    return false;
                }
            }
        }
        return stack.size() == 0;

    }

    public static void main(String[] var0) {
        System.out.println(isValid("(]"));
    }
}

File: Easy/20.Valid-Parentheses/javascript/Valid_Parentheses.js
class Stack {
  constructor() {
    this.items = []
  }
  push(el) {
    this.items.push(el)
  }
  pop() {
    return this.items.pop()
  }
  size() {
    return this.items.length
  }
}

var isValid = (s) => {
  const map = new Map()
  map.set(')', '(')
  map.set(']', '[')
  map.set('}', '{')
  const stack = new Stack()
  for (let i = 0; i < s.length; i++) {
    if (map.has(s[i])) {
      if (map.get(s[i]) !== stack.pop()) return false
    } else {
      stack.push(s[i])
    }
  }
  return stack.size() === 0
}

console.log(isValid(']'))
